File: slnSopimusrekisteriLuokat/Sopimusrekisteri.BLL_CF/Sopimus.cs
using Sopimusrekisteri.BLL_CF.Class;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sopimusrekisteri.BLL_CF
{
    public class Sopimus : IEditable, IPoimittava
    {
        #region Propertyt

        public int Id { get; set; }
        public Sopimustyypit SopimustyyppiId { get; set; }
        public int? PaasopimusId { get; set; }
        public DateTime? Alkaa { get; set; }
        public DateTime? Paattyy { get; set; }
        public int? KestoId { get; set; }
        public DateTime? ProjektiAloitusPvm { get; set; }
        public string MuuTunniste { get; set; }
        public string Iban { get; set; }
        public int? JuridinenYhtioId { get; set; }
        public string SopimuksenLaatija { get; set; }
        public int? VastuuyksikkoId { get; set; }
        public string Projektinumero { get; set; }
        public int? Pylvasmaara { get; set; }
        public bool Luonnonsuojelualue { get; set; }
        public string Pylvasvali { get; set; }
        public string MaantieteellinenVali { get; set; }
        public decimal? Muuntamoalue { get; set; }
        public int? JulkisuusasteId { get; set; }
        public string Korvaa { get; set; }
        public SopimusTilat? SopimuksenTilaId { get; set; }
        public string Info { get; set; }
        public DateTime Luotu { get; set; }
        public string Luoja { get; set; }
        public string Paivittaja { get; set; }
        public DateTime? Paivitetty { get; set; }
        public decimal? AlkuperainenKorvaus { get; set; }
        public int? DFRooliId { get; set; }
        public int? SopimusluokkaId { get; set; }
        public int? SopimuksenAlaluokkaId { get; set; }
        public int? SopimuksenEhtoversioId { get; set; }
        public string Karttaliite { get; set; }
        public string Kuvaus { get; set; }
        public DateTime? VerkonhaltijanAllekirjoitusPvm { get; set; }
        public DateTime? AsiakkaanAllekirjoitusPvm { get; set; }
        public int? SopimuksenIrtisanomisaika { get; set; }
        public string SopimuksenIrtisanomistoimet { get; set; }
        public int? VerkonhaltijaSiirtoOikeusId { get; set; }
        public int? VastaosapuoliSiirtoOikeusId { get; set; }
        public int? AlkuperainenYhtioId { get; set; }
        public DateTime? Irtisanomispvm { get; set; }
        public string PCSNumero { get; set; }
        public string Projektinvalvoja { get; set; }
        public int? PuustonPoistoId { get; set; }
        public int? PuustonOmistajuusId { get; set; }
        public int? Jatkoaika { get; set; }
        public int? KohdekategoriaId { get; set; }
        public bool TiedostoHaettu { get; set; }
        public bool MetatiedotPaivitetty { get; set; }
        public bool Luonnos { get; set; }
        public Kielet? KieliId { get; set; }
        public string Erityisehdot { get; set; }
        public int? YlasopimuksenTyyppiId { get; set; }
        public int? LupatahoId { get; set; }
        public bool Korvaukseton { get; set; }
        public string Mappitunniste { get; set; }
        public string CaceTehtava { get; set; }
        public DateTime? SopimusarkistosiirtoTehty { get; set; }
        public string Guid { get; set; }
        public int? VuokratyyppiId { get; set; }
        public decimal? Korkoprosentti { get; set; }

        /// <summary>
        /// Vain tiedoksi. Ei käytetä kyselyissä tai hauissa
        /// </summary>
        public int? LahdeKorkoprosenttiId { get; set; }

        /// <summary>
        /// Finnish Accounting Standards
        /// </summary>
        public bool FAS { get; set; }

        /// <summary>
        /// International Financial Reporting Standards
        /// </summary>
        public bool IFRS { get; set; }

        public int? Sopimusvuosi
        {
            get
            {
                if (this.Alkaa.HasValue)
                    return this.Alkaa.Value.Year;

                return null;
            }
        }

        /// <summary>
        /// Hakee laskennallisen päättymispäivän kannasta tallennettuna
        /// </summary>
        public DateTime? LaskennallinenPaattymispvm { get; set; }

        //public DateTime? LaskennallinenPaattymispvm
        //{
        //    get
        //    {
        //        if (this.Paattyy.HasValue)
        //        {

        //            DateTime paattymispvmIrtisanomisajalla = this.GetPvmIrtisanomisajalla(this.Paattyy.Value);

        //            if (paattymispvmIrtisanomisajalla < DateTime.Now.Date && paattymispvmIrtisanomisajalla < this.Irtisanomispvm.GetValueOrDefault(DateTime.MaxValue))
        //            {

        //                if (this.Jatkoaika.HasValue && this.Jatkoaika.Value > 0)
        //                {

        //                    DateTime laskennallinenPaattymispvm = this.Paattyy.Value.AddMonths(this.Jatkoaika.Value);

        //                    DateTime laskennallinenPaattymispvmIrtisanomisajalla = this.GetPvmIrtisanomisajalla(laskennallinenPaattymispvm);

        //                    while (laskennallinenPaattymispvmIrtisanomisajalla < DateTime.Now.Date && laskennallinenPaattymispvmIrtisanomisajalla < this.Irtisanomispvm.GetValueOrDefault(DateTime.MaxValue))
        //                    {

        //                        laskennallinenPaattymispvm = laskennallinenPaattymispvm.AddMonths(this.Jatkoaika.Value);
        //                        laskennallinenPaattymispvmIrtisanomisajalla = this.GetPvmIrtisanomisajalla(laskennallinenPaattymispvm);

        //                    }

        //                    return laskennallinenPaattymispvm;

        //                }

        //            }

        //            return this.Paattyy.Value;

        //        }

        //        return null;

        //    }
        //}

        private DateTime GetPvmIrtisanomisajalla(DateTime pvm)
        {
            return pvm.AddMonths(-1 * this.SopimuksenIrtisanomisaika.GetValueOrDefault(0));
        }

        /// <summary>
        /// Laskee laskennallisen päättymispäivät muista tiedoista
        /// </summary>
        public LaskennallinenPaattymispvm DynaaminenLaskennallinenPaattymispvm => new LaskennallinenPaattymispvm
        {
            Irtisanomisaika = SopimuksenIrtisanomisaika,
            Irtisanottu = Irtisanomispvm,
            Paattyy = Paattyy,
            Jatkoaika = Jatkoaika,
            Alkaa = Alkaa
        };

        public string Nimi
        {
            get { return this.Id.ToString() + string.Format("({0})", this.MuuTunniste); }
        }

        public string SopimustyypinNimi
        {
            get
            {
                if (this.YlasopimuksenTyyppi != null)
                    return this.YlasopimuksenTyyppi.Nimi;

                if (this.Sopimustyyppi != null)
                    return this.Sopimustyyppi.SopimustyyppiNimi;

                return string.Empty;
            }
        }

        public bool Hyvaksytty
        {
            get
            {
                return this.AsiakkaanAllekirjoitusPvm.HasValue;
            }
        }

        public IEnumerable<string> Kiinteistotunnukset
        {
            get
            {
                if (this.Kiinteistot != null)
                    return this.Kiinteistot.Where(x => !string.IsNullOrEmpty(x.KiinteistotunnusLyhyt)).Select(x => x.KiinteistotunnusLyhyt).Distinct();

                return null;
            }
        }

        public IEnumerable<string> Kunnat
        {
            get
            {
                if (this.Kiinteistot != null)
                    return this.Kiinteistot.Where(x => x.Kunta != null).Where(x => !string.IsNullOrEmpty(x.Kunta.KuntaNimi)).Select(x => x.Kunta.KuntaNimi).Distinct();

                return null;
            }
        }

        public IEnumerable<string> Kylat
        {
            get
            {
                if (this.Kiinteistot != null)
                    return this.Kiinteistot.Where(x => !string.IsNullOrEmpty(x.Kyla)).Select(x => x.Kyla).Distinct();

                return null;
            }
        }

        public IEnumerable<string> Rekisterinumerot
        {
            get
            {
                if (this.Kiinteistot != null)
                    return this.Kiinteistot.Where(x => !string.IsNullOrEmpty(x.Rekisterinumero)).Select(x => x.Rekisterinumero).Distinct();

                return null;
            }
        }

        public IEnumerable<string> Tilat
        {
            get
            {
                if (this.Kiinteistot != null)
                    return this.Kiinteistot.Where(x => !string.IsNullOrEmpty(x.KiinteistoNimi)).Select(x => x.KiinteistoNimi).Distinct();

                return null;
            }
        }

        public IEnumerable<string> Kohteet
        {
            get
            {
                if (this.Tunnisteyksikot != null)
                {
                    List<string> kohteet = new List<string>();

                    kohteet.AddRange(this.Tunnisteyksikot.Where(x => !string.IsNullOrEmpty(x.PGTunnus)).Select(x => x.PGTunnus).Distinct());
                    kohteet.AddRange(this.Tunnisteyksikot.Where(x => string.IsNullOrEmpty(x.PGTunnus) && !string.IsNullOrEmpty(x.Nimi)).Select(x => x.Nimi).Distinct());

                    return kohteet;
                }

                return null;
            }
        }

        public IEnumerable<string> Sopimusosapuolet
        {
            get
            {
                if (this.Asiakkaat != null)
                    return this.Asiakkaat.Where(x => x.Taho != null && !string.IsNullOrEmpty(x.Taho.Nimi)).Select(x => x.Taho.Nimi);

                return null;
            }
        }

        public IEnumerable<string> AlkuperaisetOsapuolet
        {
            get
            {
                if (this.Asiakkaat != null)
                    return this.Asiakkaat.Where(x => x.AsiakastyyppiId == (int)Asiakastyypit.AlkuperainenOsapuoli && x.Taho != null && !string.IsNullOrEmpty(x.Taho.Nimi)).Select(x => x.Taho.Nimi);

                return null;
            }
        }

        #endregion

        #region Entiteettiviittaukset

        public virtual Sopimustyyppi Sopimustyyppi { get; set; }
        public virtual Sopimus Paasopimus { get; set; }
        public virtual SopimuksenKesto Kesto { get; set; }
        public virtual Taho JuridinenYhtio { get; set; }
        public virtual Taho Vastuuyksikko { get; set; }
        public virtual Julkisuusaste Julkisuusaste { get; set; }
        public virtual SopimuksenTila SopimuksenTila { get; set; }
        public virtual DFRooli DFRooli { get; set; }
        public virtual Sopimusluokka Sopimusluokka { get; set; }
        public virtual SopimuksenAlaluokka SopimuksenAlaluokka { get; set; }
        public virtual SopimuksenEhtoversio SopimuksenEhtoversio { get; set; }
        public virtual SiirtoOikeus VerkonhaltijaSiirtoOikeus { get; set; }
        public virtual SiirtoOikeus VastaosapuoliSiirtoOikeus { get; set; }
        public virtual Taho AlkuperainenYhtio { get; set; }
        public virtual PuustonPoisto PuustonPoisto { get; set; }
        public virtual PuustonOmistajuus PuustonOmistajuus { get; set; }
        public virtual Kohdekategoria Kohdekategoria { get; set; }
        public virtual Kieli Kieli { get; set; }
        public virtual YlasopimuksenTyyppi YlasopimuksenTyyppi { get; set; }
        public virtual Lupataho Lupataho { get; set; }
        public virtual Vuokratyyppi Vuokratyyppi { get; set; }

        public virtual ICollection<SopimusKiinteisto> SopimusKiinteistot { get; set; }
        public virtual ICollection<SopimusTaho> Asiakkaat { get; set; }
        public virtual ICollection<Tunnisteyksikko> Tunnisteyksikot { get; set; }
        public virtual ICollection<Tiedosto> Tiedostot { get; set; }
        public virtual ICollection<Korvauslaskelma> Korvauslaskelmat { get; set; }

        public IEnumerable<Kiinteisto> Kiinteistot
        {
            get
            {
                if (this.SopimusKiinteistot != null)
                    return this.SopimusKiinteistot.Select(x => x.Kiinteisto);

                return null;
            }
        }

        #endregion

        #region Metodit

        public override bool Equals(object obj)
        {
            var tmp = obj as Sopimus;
            if (tmp == null) return false;
            return tmp.Id == Id;
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }

        public override string ToString()
        {
            return Id.ToString();
        }

        #endregion
    }
}

File: appSopimusrekisteri.DTO/Sopimus.vb
Public Class Sopimus

    Private _strSopimusVuosi As String
    Private _dAlkupvm As Date?

    Public Sub New()
        Me.Kiinteistot = New List(Of Kiinteisto)()
        Me.Tahot = New List(Of SopimusTaho)()
        Me.Tunnisteyksikot = New List(Of Tunnisteyksikko)()
        Me.Tiedostot = New List(Of Tiedosto)()
        Me.Korvauslaskelmat = New List(Of Korvauslaskelma)()
    End Sub

    Public Property Id As Integer
    Public Property Sopimustyyppi As String
    Public Property Tyyppi As String
    Public Property Alkaa As String
    Public Property Paattyy As String

    Public Property MuuTunniste As String
    Public Property Vuosi As String
    'Vain ensimmäisen kiinteistön tiedot näytetään tällä hetkellä. Entä sopimukset joissa on monta kiinteistöä?
    Public ReadOnly Property KiinteistonKunta As String
        Get
            If Kiinteistot.Count > 0 Then Return Kiinteistot.First.Kunta Else Return ""
        End Get
    End Property
    Public ReadOnly Property KiinteistonKyla As String
        Get
            If Kiinteistot.Count > 0 Then Return Kiinteistot.First.Kyla Else Return ""
        End Get
    End Property
    Public ReadOnly Property KiinteistonRekisterinumero As String
        Get
            If Kiinteistot.Count > 0 Then Return Kiinteistot.First.Rekisterinumero Else Return ""
        End Get
    End Property
    Public ReadOnly Property KiinteistonTunnusLyhyt As String
        Get
            If Kiinteistot.Count > 0 Then Return Kiinteistot.First.LyhytKiinteistotunnus Else Return ""
        End Get
    End Property
    Public Property Asiakkaat As String
    Public Property PGTunnus As String
    Public Property Lisatieto As String
    Public Property PCSNumero As String

    Public Property Alkupvm As Date?
        Set(value As Date?)
            _dAlkupvm = value
        End Set
        Get
            If _dAlkupvm.HasValue Then
                Return _dAlkupvm
            End If

            Return Me.VerkonhaltijanAllekirjoitusPvm
        End Get
    End Property


    Public Property Irtisanomispvm As Date?
    Public Property Paattymispvm As Date?

    Public Property Korvaukseton As Boolean

    Public Property TiedostoHaettu As Boolean = False
    Public Property MetatiedotPaivitetty As Boolean = False

    Public Property Luoja As String
    Public Property Luotu As Date?
    Public Property Paivittaja As String
    Public Property Paivitetty As Date?

    Public Property AlkuperainenKorvaus As Decimal?
    Public Property AsiakkaanAllekirjoitusPvm As Date?
    Public Property DFRooliId As Integer?
    Public Property DFRooli As String
    Public Property Info As String
    Public Property JulkisuusasteId As Integer?
    Public Property Julkisuusaste As String
    Public Property JuridinenYhtioId As Integer?
    Public Property Karttaliite As String
    Public Property KestoId As Integer?
    Public Property Korvaa As String
    Public Property Kuvaus As String
    Public Property Luonnonsuojelualue As Boolean = False
    Public Property MaantieteellinenVali As String
    Public Property Muuntamoalue As Decimal?
    Public Property PaasopimusId As Integer?
    Public Property Paasopimus As Sopimus
    Public Property ProjektiAloitusPvm As Date?
    Public Property Pylvasmaara As Integer?
    Public Property Pylvasvali As String
    Public Property SopimuksenAlaluokkaId As Integer?
    Public Property SopimuksenAlaluokka As String
    Public Property SopimuksenEhtoversioId As Integer?
    Public Property SopimuksenEhtoversio As String
    Public Property SopimuksenIrtisanomisaika As Integer?
    Public Property SopimuksenIrtisanomistoimet As String
    Public Property SopimuksenLaatija As String
    Public Property SopimuksenTilaId As Integer?
    Public Property SopimuksenTila As String
    Public Property SopimusluokkaId As Integer?
    Public Property Sopimusluokka As String
    Public Property SopimustyyppiId As Integer?
    Public Property VastaosapuoliSiirtoOikeusId As Integer?
    Public Property VastaosapuoliSiirtoOikeus As String
    Public Property VastuuyksikkoId As Integer?
    Public Property Vastuuyksikko As String
    Public Property VerkonhaltijaSiirtoOikeusId As Integer?
    Public Property VerkonhaltijaSiirtoOikeus As String
    Public Property VerkonhaltijanAllekirjoitusPvm As Date?
    Public Property AlkuperainenYhtioId As Integer?
    Public Property AlkuperainenYhtio As String
    Public Property Projektinvalvoja As String
    Public Property Jatkoaika As Integer?
    Public Property PuustonOmistajuusId As Integer?
    Public Property PuustonOmistajuus As String
    Public Property PuustonPoistoId As Integer?
    Public Property PuustonPoisto As String
    Public Property KohdekategoriaId As Integer?
    Public Property Kohdekategoria As String
    Public Property LaskennallinenPaattymispvm As Date?

    Public Property Luonnos As Boolean = False
    Public Property KieliId As Integer?
    Public Property Kieli As String
    Public Property Erityisehdot As String

    Public Property YlasopimuksenTyyppiId As Integer?
    Public Property YlasopimuksenTyyppi As String

    Public Property LupatahoId As Integer?
    Public Property Lupataho As String

    Public Property Kiinteistot As List(Of Kiinteisto)
    Public Property Tahot As List(Of SopimusTaho)
    Public Property Tunnisteyksikot As List(Of Tunnisteyksikko)
    Public Property Korvauslaskelmat As List(Of Korvauslaskelma)
    Public Property Tiedostot As List(Of Tiedosto)

    Public Property JuridinenYhtio As Taho
    Public Property Mappitunniste As String
    Public Property CaceTehtava As String

    Public ReadOnly Property JuridinenYhtioNimi As String
        Get
            If Not Me.JuridinenYhtio Is Nothing Then
                Return Me.JuridinenYhtio.Nimi
            Else
                Return String.Empty
            End If
        End Get
    End Property

    Public ReadOnly Property Hyvaksytty As Boolean
        Get
            If Me.AsiakkaanAllekirjoitusPvm.HasValue Then
                Return True
            Else
                Return False
            End If
        End Get
    End Property

    'Public ReadOnly Property LaskennallinenPaattymispvm As Date?
    '  Get
    '    If Me.Paattymispvm.HasValue Then

    '      Dim intIrtisanomisaika As Integer = Me.SopimuksenIrtisanomisaika.GetValueOrDefault(0)
    '      Dim pvmIrtisanomisajalla As Date = DateAdd(DateInterval.Month, -1 * intIrtisanomisaika, Me.Paattymispvm.Value)

    '      If pvmIrtisanomisajalla < Date.Now And pvmIrtisanomisajalla < Me.Irtisanomispvm.GetValueOrDefault(Date.MaxValue) Then

    '        If Me.Jatkoaika.HasValue AndAlso Me.Jatkoaika.Value > 0 Then

    '          Dim pvm As Date = DateAdd(DateInterval.Month, Me.Jatkoaika.Value, Me.Paattymispvm.Value)
    '          pvmIrtisanomisajalla = DateAdd(DateInterval.Month, -1 * intIrtisanomisaika, pvm)

    '          While pvmIrtisanomisajalla < Date.Now And pvmIrtisanomisajalla < Me.Irtisanomispvm.GetValueOrDefault(Date.MaxValue)
    '            pvm = DateAdd(DateInterval.Month, Me.Jatkoaika.Value, pvm)
    '            pvmIrtisanomisajalla = DateAdd(DateInterval.Month, -1 * intIrtisanomisaika, pvm)
    '          End While

    '          Return pvm
    '        End If

    '      End If

    '      Return Me.Paattymispvm

    '    End If

    '    Return Nothing
    '  End Get
    'End Property

    Public Property Sopimusvuosi As String
        Set(value As String)
            _strSopimusVuosi = value
        End Set
        Get
            If Me.Alkupvm.HasValue Then
                Return Year(Me.Alkupvm).ToString()
            Else
                Return _strSopimusVuosi
            End If
        End Get
    End Property

    Public ReadOnly Property Kiinteistotunnukset As String()
        Get
            If Not Me.Kiinteistot Is Nothing Then
                Return Me.Kiinteistot.Where(Function(x) Not String.IsNullOrEmpty(x.LyhytKiinteistotunnus)).Select(Function(x) x.LyhytKiinteistotunnus).Distinct().ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Kunnat As String()
        Get
            If Not Me.Kiinteistot Is Nothing Then
                Return Me.Kiinteistot.Where(Function(x) Not String.IsNullOrEmpty(x.Kunta)).Select(Function(x) x.Kunta).Distinct().ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Kylat As String()
        Get
            If Not Me.Kiinteistot Is Nothing Then
                Return Me.Kiinteistot.Where(Function(x) Not String.IsNullOrEmpty(x.Kyla)).Select(Function(x) x.Kyla).Distinct().ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Rekisterinumerot As String()
        Get
            If Not Me.Kiinteistot Is Nothing Then
                Return Me.Kiinteistot.Where(Function(x) Not String.IsNullOrEmpty(x.Rekisterinumero)).Select(Function(x) x.Rekisterinumero).Distinct().ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Tilat As String()
        Get
            If Not Me.Kiinteistot Is Nothing Then
                Return Me.Kiinteistot.Where(Function(x) Not String.IsNullOrEmpty(x.Nimi)).Select(Function(x) x.Nimi).Distinct().ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Sopimusosapuolet As String()
        Get
            If Not Me.Tahot Is Nothing Then
                Return Me.Tahot.Select(Function(x) x.Taho).Where(Function(x) Not String.IsNullOrEmpty(x.Nimi)).Select(Function(x) x.Nimi).Distinct.ToArray()
            Else
                Return Nothing
            End If
        End Get
    End Property

    Public ReadOnly Property Kohteet As String()
        Get
            Dim lstKohteet As New List(Of String)()
            If Not Me.Tunnisteyksikot Is Nothing Then

                lstKohteet.AddRange(Me.Tunnisteyksikot.Where(Function(x) Not String.IsNullOrEmpty(x.PGTunnus)).Select(Function(x) x.PGTunnus).Distinct)
                lstKohteet.AddRange(Me.Tunnisteyksikot.Where(Function(x) String.IsNullOrEmpty(x.PGTunnus) And Not String.IsNullOrEmpty(x.Nimi)).Select(Function(x) x.Nimi).Distinct)

                If lstKohteet.Count > 0 Then
                    Return lstKohteet.ToArray()
                End If

            End If
            Return Nothing
        End Get
    End Property

    Public ReadOnly Property Nimi
        Get
            Return Me.Id.ToString() & " " & String.Format("({0})", LTrim(Me.MuuTunniste & " " & Me.Vuosi))
        End Get
    End Property

End Class
